riptide-compression/src/test/java/org/zalando/riptide/compression/RequestCompressionPluginTest.java

```diff
@@ -0,0 +1,120 @@
+        executor.awaitTermination(10, TimeUnit.SECONDS);
+        driver.verify();
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(RequestFactorySource.class)
+    void shouldCompressRequestUsingFactory(final ClientHttpRequestFactory factory) {
+        driver.addExpectation(onRequestTo("/")
+                        .withMethod(POST)
+                        .withHeader("X-Content-Encoding", "gzip") // written by Jetty's GzipHandler
+                        .withBody(equalTo("{}"), "application/json"),
+                giveResponse("", "text/plain"));
+
+        final Http http = buildHttp(factory, new RequestCompressionPlugin());
+        http.post("/")
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(new HashMap<>())
+                .call(pass())
+                .join();
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(RequestFactorySource.class)
+    void shouldNotCompressEmptyBody(final ClientHttpRequestFactory factory) {
+        driver.addExpectation(onRequestTo("/")
+                        .withBody(emptyString(), "text/plain")
+                        .withoutHeader("Content-Encoding")
+                        .withoutHeader("X-Content-Encoding"),
+                giveResponse("", "text/plain"));
+
+        final Http http = buildHttp(factory, new RequestCompressionPlugin());
+        http.get("/")
+                .contentType(MediaType.TEXT_PLAIN)
+                .call(pass())
+                .join();
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(RequestFactorySource.class)
+    void shouldCompressWithGivenAlgorithm(final ClientHttpRequestFactory factory) {
+        driver.addExpectation(onRequestTo("/")
+                        .withMethod(POST)
+                        .withHeader("Content-Encoding", "identity") // not handled by Jetty
+                        .withoutHeader("X-Content-Encoding")
+                        .withBody(equalTo("{}"), "application/json"),
+                giveResponse("", "text/plain"));
+
+        final Http http = buildHttp(factory, new RequestCompressionPlugin(Compression.of("identity", it -> it)));
+        http.post("/")
+                .contentType(MediaType.APPLICATION_JSON)
+                .body(new HashMap<>())
+                .call(pass())
+                .join();
+    }
+
+    private Http buildHttp(final ClientHttpRequestFactory factory, final Plugin... plugins) {
+        return Http.builder()
+                .executor(executor)
+                .requestFactory(factory)
+                .baseUrl(driver.getBaseUrl())
+                .plugins(asList(plugins))
+                .build();
+    }
+
+    static class RequestFactorySource implements ArgumentsProvider {
+        @Override
+        public Stream<? extends Arguments> provideArguments(ExtensionContext extensionContext) {
+            return Stream.of(
+                    new SimpleClientHttpRequestFactory(),
+                    new Netty4ClientHttpRequestFactory(),
```

Comment:
Starting Spring 4.3 the `Netty4ClientHttpRequest` unconditionally adds a `Content-Length` header, see [here](https://github.com/spring-projects/spring-framework/blob/master/spring-web/src/main/java/org/springframework/http/client/Netty4ClientHttpRequest.java#L144-L146). Therefore servers are rejecting the request (clashes with the `chunked` encoding).

@whiskeysierra I propose we document this in the README telling users to not use Netty if they want the compression plugin. Wdyt?

Reply:
Adapted.